feat: Support for Alternative Items in Quotation

marination opened the pull request:
> **Note:** Not to be confused with DocType "Item Alternative"

Documentation Draft: https://docs.erpnext.com/docs/v14/user/manual/en/selling/quotation/edit-wiki?wiki_page_patch=b1a88d79a9

Please check https://github.com/frappe/erpnext/issues/33609 to understand the issue

<details>
<summary>Impact on Quotation</summary>

> **Note:** It is important to maintain the right order i.e. Alternative Item rows must follow a Non-alternative item row. Grouping will be done on this basis
> <img width="350" alt="Screenshot 2023-02-06 at 4 38 13 PM" src="https://user-images.githubusercontent.com/25857446/216957147-a47d4c5c-4390-477a-9abb-a7d0ed3bb76e.png">


- Users can add an alternative item in the Quotation to give the customer a chance to choose among items in the quotation
- To add an alternative item: add a row and check **Is Alternative** 
    <img width="966" alt="Screenshot 2023-02-06 at 4 31 18 PM" src="https://user-images.githubusercontent.com/25857446/216955223-6f564c53-3abf-4455-973d-a035f254bc06.png">

- **Totals**: Only consider original (non-alternative Item)
  <img width="1039" alt="Screenshot 2023-01-31 at 5 02 44 PM" src="https://user-images.githubusercontent.com/25857446/215748953-8f4846d1-12cd-4617-8216-1c0c89ec5431.png">
- **Taxes**: Applied accordingly
  <img width="1034" alt="Screenshot 2023-01-31 at 5 03 45 PM" src="https://user-images.githubusercontent.com/25857446/215749139-91e4ca04-4aa7-4fcc-9d7e-9d91afaf7ecc.png">
</details>

<details>
<summary>Mapping to Sales Order</summary>
Consider:
<img width="1026" alt="Screenshot 2023-01-31 at 5 05 31 PM" src="https://user-images.githubusercontent.com/25857446/215749415-0d1c179e-8543-41af-a1f4-755f9811bd37.png">

- User will be asked to select an item from among the original-alternatives set, to map in the Sales Order
   ![2023-02-06 16 44 59](https://user-images.githubusercontent.com/25857446/216957964-021ceb1c-48b7-4166-a5bf-c39f45e936ef.gif)

- One of the alternatives is selected, the same is mapped and the rest in the set are skipped. Apart from that the simple items are also mapped:
   ![2023-02-06 16 46 16](https://user-images.githubusercontent.com/25857446/216958157-64e9fab2-d45e-4bab-b79f-c7ebe42c6961.gif)
</details>

<details>
<summary>Use Cases</summary>

- Item Code has alternatives which are different Items
   <img width="1014" alt="Screenshot 2023-02-06 at 4 51 28 PM" src="https://user-images.githubusercontent.com/25857446/216959057-4718ee10-58e1-4167-9c5e-71bb6b6081a7.png">
- Service Item has alternatives and non-alternatives sharing the same Item code. The case here is that the service costing is very dynamic and subjective and so the **same code** with different Item name/Description is used in the Quotation.
 <img width="1027" alt="Screenshot 2023-02-06 at 4 53 46 PM" src="https://user-images.githubusercontent.com/25857446/216959394-6ccc4ee2-b412-4a5b-b3b5-5e2ff7e194e8.png">

</details>

### Todo

- [x]  Partially ordered status
- [x] Tests

marination committed:
feat: Filter out alternative item rows in taxes and totals for Quotation

- Added a Quotation Item field `is_alternative_item`
- Use filtered rows for taxes and totals computation

marination committed:
feat: Consider filtered items table in JS for totals computation

- Set `_items` as filtered rows if quotation else the entire table. Set at entry point of JS API
- Use `_items` instead of `items` to compute taxes and charges. Exclude alternative item rows

marination committed:
feat: Dialog to select alternative item before creating Sales order

- Users can leave the row blank in the dialog if original item is to be used
- Else users can select an alternative item against an original item
- In the document, users must check `Is Alternative Item` if needed and also specify which item it is an altenrative to since there are no documented mappings

marination committed:
feat: Filter rows to be mapped on server side mapping function

- Pass dialog selections to `make_sales_order`
- Map either original item or its alternative depending on mapping
- Only qty check for simple rows (without alternatives and not an alternative itself)

marination committed:
chore: Validate 'alternative_to' field values, must be a valid non-alterntaive item from table

marination committed:
fix: Iterate over list instead of map's output and formatting

marination committed:
fix: Consider only ordered alternative/original item for Quotation status

- The original and its alternatives make a set of items where one is chosen
- While setting order status of Quotation, check if the chosen item from the set is fully ordered or not
- Filter out unselected items from the set
- Create a map containing the set of items and if they were ordered or not for ease of grouping
- The simple items will work as it used to

marination committed:
chore: Code simplification

- Map is not required, avoid filter multiple times, use single loop instead
- Better variable name
- Reduce LOC

marination committed:
refactor: Order based alternative items mapping

- Alternatives must be followed by a non-alternative item row
- On submit, store non-alternative rows in hidden checkbox to avoid recomputation
- Check for valid/mappable rows by row name
- UI: Select from table rows.Add single row for original/alternative item in dialog
- UI: Indicator for alternative items in dialog grid
- UI: Indicator legend and description of table
- DB: Added check field 'Has Alternative Item' not to be confused with 'Has Alternative' in Mfg

marination committed:
Merge branch 'develop' into alternative-items-quotation

marination committed:
Merge branch 'develop' into alternative-items-quotation

marination committed:
test: Alternative items in Quotation

- Taxes and totals, mapping, back updation

deepeshgarg007 commented on erpnext/selling/doctype/quotation/quotation.js:
```diff
@@ -220,17 +234,111 @@ erpnext.selling.QuotationController = class QuotationController extends erpnext.
 	}
+
+	show_alternative_items_dialog() {
+		var me = this;
```
```suggestion
		let me = this;
```

deepeshgarg007 reviewed (commented):
The dialog won't work if someone uses the "Get Items From" option directly from the sales order, right?

marination commented:
@deepeshgarg007 Yes it won't. I'll fix that

marination committed:
fix: Use block variable

Co-authored-by: Deepesh Garg <deepeshgarg6@gmail.com>

marination committed:
fix: Handle `Get Items From` in Sales Order

- Map all non alternatives from Quotation to SO if no selected items
- Show disclaimer mentioning that Qtns with alternatives must be mapped to SO from the Qtn form

marination committed:
fix: Map only non alternative items from Quotation in Sales Invoice

- Since there's no item selection, only Quotation selection :/

marination commented:
@deepeshgarg007 in SO/SI item selection is not possible/doesn't make sense, because i don't think any user wants to select alternatives from a mountain of rows from multiple Quotations.

So i added a disclaimer in the SO dialog saying that if alternatives are involved we would recommend mapping from the Quotation. And then if the user proceeds anyway, only non alternative rows are mapped in the Sales Order (the rows that sum up to the total in the quotation)
<img width="1199" alt="Screenshot 2023-02-20 at 9 21 19 PM" src="https://user-images.githubusercontent.com/25857446/220152404-264dd5d8-4acb-4f26-afd1-b4a950b47ebb.png">

marination committed:
Merge branch 'develop' into alternative-items-quotation

maveezmh commented:
@marination how about having a provision where user can deliver an alternative because given item is not aviailable in stock
